keywrap/pkcs11/keywrapper_pkcs11.go

```diff
@@ -41,7 +41,11 @@ func NewKeyWrapper() keywrap.KeyWrapper {
 // WrapKeys wraps the session key for recpients and encrypts the optsData, which
 // describe the symmetric key used for encrypting the layer
 func (kw *pkcs11KeyWrapper) WrapKeys(ec *config.EncryptConfig, optsData []byte) ([]byte, error) {
-	pkcs11Recipients, err := addPubKeys(&ec.DecryptConfig, append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-yamls"]...))
+	// append({}, ...) allocates a fresh backing array, and that's necessary to guarantee concurrent calls to WrapKeys (as in c/image/copy.Image)
+	// can't race writing to the same backing array.
+	pubKeys := append([][]byte{}, ec.Parameters["pkcs11-pubkeys"]...) // In Go 1.21, slices.Clone(ec.Parameters["pkcs11-pubkeys"])
+	pubKeys = append(pubKeys, ec.Parameters["pkcs11-yamls"]...)
```

Comment:
Regarding the patch description: This would sound like local variables were raceable by other co-routines. Can this be?

Reply:
`ec` is not a local variable; it is provided to `copy.Image`, which spins up (typically) 6 goroutines concurrently copying image layers. Therefore `ec.Parameters` is possibly concurrently read by those goroutines, and in the current version of the code, `append(ec.Parameters[…]` is possibly concurrently called on the same slice.

From a quick search, `append` amounts to probaly something like https://github.com/golang/go/blob/2968f5623c71d0e8e7be6e117939c5f585bb0620/src/cmd/compile/internal/walk/builtin.go#L43 , calling https://github.com/golang/go/blob/2968f5623c71d0e8e7be6e117939c5f585bb0620/src/runtime/slice.go#L174 , which (to avoid an O(N^2) behavior on repeated appends), can create a slice with `cap(slice) > len(slice)`. And with such a slice, `append(slice_with_large_cap, new_elem)` is basically `slice_with_large_cap[len++] = new_elem` — a concurrent write to memory shared across goroutines.

---

The fix forces creation of a new slice, local to this function.

Reply:
How is the slice shared across go-routines? It's `ec` that in your case is shared across go-routines, but the slice should be local to each go-routine, no?

Reply:
`ec` is shared, and potentially concurrently read from multiple goroutines
→ they all share, and potentially concurrently read, the `ec.Parameters` map
→ they all share, and potentially concurrently read, the `ec.Parameters["pkcs11-pubkeys"]` slice value
→ they all share the _array_ underlying that slice
→ they all potentially write to that backing array.

Reply:
In your proposed solution you are now creating an explicit slice `pubKeys` that presumably was done implicitly also before in `append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-yamls"]...)`.

Per what you are saying above the `ec` is shared and 'they all potentially write to that backing array'. They all write to the backing array of their own private slice 'pubKeys', right?

Reply:
@lumjjb Comments?

Reply:
I could see a race if they were writing to a shared `pubKeys` slice but I don't see that happening. Nor do I see races due to reading from `ec`.

Reply:
> In your proposed solution you are now creating an explicit slice `pubKeys` that presumably was done implicitly also before in `append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-yamls"]...)`.

No. `append` modifies the backing array of the first parameter. It does not _always_ create a new array.

Reply:
> > In your proposed solution you are now creating an explicit slice `pubKeys` that presumably was done implicitly also before in `append(ec.Parameters["pkcs11-pubkeys"], ec.Parameters["pkcs11-yamls"]...)`.
> 
> No. `append` modifies the backing array of the first parameter. It does not _always_ create a new array.

But each coroutine has its own slice. So how would one coroutine influence another one?

Reply:
Its own slice, not its own backing array. Each slice is individually grown, but they all overwrite the previously-free slots in the array.